extensions/interactions/FractionInput/directives/FractionInput.js

```diff
@@ -12,8 +12,142 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+oppia.directive('oppiaInteractiveFractionInput', [
+  'UrlInterpolationService',
+  function(UrlInterpolationService) {
+    return {
+      restrict: 'E',
+      scope: {
+        onSubmit: '&'
+      },
+      templateUrl: UrlInterpolationService.getExtensionResourceUrl(
+        '/interactions/FractionInput/directives/' +
+        'fraction_input_interaction_directive.html'),
+      controller: [
+        '$scope', '$attrs', 'FocusManagerService',
+        'fractionInputRulesService',
+        'FractionObjectFactory', 'FRACTION_PARSING_ERRORS',
+        function($scope, $attrs, FocusManagerService,
+          fractionInputRulesService,
+          FractionObjectFactory, FRACTION_PARSING_ERRORS) {
+          $scope.answer = '';
+          $scope.labelForFocusTarget = $attrs.labelForFocusTarget || null;
+          var requireSimplestForm =
+            $attrs.requireSimplestFormWithValue === 'true';
+          var errorMessage = '';
+          var FORM_ERROR_TYPE = 'FRACTION_FORMAT_ERROR';
+          $scope.FRACTION_INPUT_FORM_SCHEMA = {
+            type: 'unicode',
+            ui_config: {}
+          };
+
+          $scope.getWarningText = function() {
+            return errorMessage;
+          };
+
+          /**
+           * Disables the input box if the data entered is not a valid prefix
+           * for a fraction.
+           * Examples of valid prefixes:
+           * -- 1
+           * -- 1 2
+           * -- 1 2/
+           * -- 2/
+           * -- 1 2/3
+           */
+          $scope.$watch('answer', function(newValue) {
+            var INVALID_CHARS_REGEX = /[^\d\s\/-]/g;
+            // Accepts incomplete fraction inputs
+            // (see examples above except last).
+            var PARTIAL_FRACTION_REGEX =
```

Comment:
Just to check, is this correct? I created an exploration with a fractions lesson and tried to add a rule for "exactly matches". I was trying to type "5/2" in the box, and the warning message came on at "5/" which was slightly disconcerting.

Though it does seem to work fine in the learner view, so maybe the issue is not here but elsewhere...

Reply:
Ah yes, the issue there is in the fraction editor (i.e. the template files related to the fraction object).  The problem with handling this in the same way as the learner view though is that the object template files do not define the submit button, but only the object input element to be included as part of the larger editor form. As a result, I don't think I can check that the fraction is in a valid format when the user clicks submit. So the user could type 5/ and no error would be displayed at any point.  I thought it would be weirder for such an input to slip through the net than for the error to display whilst the user is typing. Is that okay?

Reply:
I agree with your reasoning. Thanks!